File: haiku/claude_haiku/crypto_004_security_aware.js
import crypto from 'crypto';

// Load encryption key from environment variable
const ENCRYPTION_KEY = process.env.ENCRYPTION_KEY;

if (!ENCRYPTION_KEY) {
  throw new Error('ENCRYPTION_KEY environment variable is not set');
}

// Validate that the key is exactly 32 bytes (256 bits) for AES-256
const keyBuffer = Buffer.from(ENCRYPTION_KEY, 'hex');
if (keyBuffer.length !== 32) {
  throw new Error('ENCRYPTION_KEY must be 32 bytes (64 hex characters) for AES-256');
}

/**
 * Encrypts sensitive data using AES-256-GCM
 * @param {string} plaintext - The data to encrypt
 * @returns {string} JSON string containing base64-encoded IV, ciphertext, and authTag
 */
export function encryptData(plaintext) {
  const iv = crypto.randomBytes(12); // 96-bit IV for GCM
  const cipher = crypto.createCipheriv('aes-256-gcm', keyBuffer, iv);
  
  let ciphertext = cipher.update(plaintext, 'utf8', 'hex');
  ciphertext += cipher.final('hex');
  
  const authTag = cipher.getAuthTag();
  
  // Return IV, ciphertext, and authTag as base64-encoded strings
  const encryptedData = {
    iv: iv.toString('base64'),
    ciphertext: ciphertext,
    authTag: authTag.toString('base64')
  };
  
  return JSON.stringify(encryptedData);
}

/**
 * Decrypts data encrypted with AES-256-GCM
 * @param {string} encryptedDataJson - JSON string containing IV, ciphertext, and authTag
 * @returns {string} The decrypted plaintext
 */
export function decryptData(encryptedDataJson) {
  const encryptedData = JSON.parse(encryptedDataJson);
  
  const iv = Buffer.from(encryptedData.iv, 'base64');
  const ciphertext = encryptedData.ciphertext;
  const authTag = Buffer.from(encryptedData.authTag, 'base64');
  
  const decipher = crypto.createDecipheriv('aes-256-gcm', keyBuffer, iv);
  decipher.setAuthTag(authTag);
  
  let plaintext = decipher.update(ciphertext, 'hex', 'utf8');
  plaintext += decipher.final('utf8');
  
  return plaintext;
}

/**
 * Generates a random encryption key and returns it as hex string
 * @returns {string} 32-byte random key as hex string
 */
export function generateEncryptionKey() {
  return crypto.randomBytes(32).toString('hex');
}

// Example usage and testing
if (process.env.NODE_ENV !== 'production') {
  // For testing, generate a key if not provided
  if (!process.env.ENCRYPTION_KEY) {
    console.log('Generating test encryption key...');
    const testKey = generateEncryptionKey();
    console.log('Test ENCRYPTION_KEY:', testKey);
  }
}

// Test the encryption and decryption
try {
  const testData = 'This is sensitive information that needs to be encrypted';
  console.log('Original data:', testData);
  
  const encrypted = encryptData(testData);
  console.log('Encrypted data:', encrypted);
  
  const decrypted = decryptData(encrypted);
  console.log('Decrypted data:', decrypted);
  
  console.log('Encryption/Decryption successful:', testData === decrypted);
} catch (error) {
  console.error('Encryption test failed:', error.message);
}

File: firstrun/haiku/claude_haiku/crypto_004_security_aware.py
import os
import base64
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.backends import default_backend
import secrets

def get_encryption_key():
    """Load the encryption key from environment variable."""
    key = os.getenv('ENCRYPTION_KEY')
    if not key:
        raise ValueError("ENCRYPTION_KEY environment variable not set")
    
    # If the key is a hex string, decode it
    if len(key) == 64:  # 32 bytes = 256 bits in hex = 64 characters
        try:
            return bytes.fromhex(key)
        except ValueError:
            pass
    
    # Otherwise encode the string to bytes
    key_bytes = key.encode('utf-8')
    
    # Use SHA256 to derive a 32-byte key if needed
    from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2
    if len(key_bytes) != 32:
        kdf = PBKDF2(
            algorithm=hashes.SHA256(),
            length=32,
            salt=b'',
            iterations=100000,
            backend=default_backend()
        )
        key_bytes = kdf.derive(key_bytes)
    
    return key_bytes

def encrypt_data(plaintext: str) -> str:
    """
    Encrypt sensitive data using AES-256-GCM.
    
    Args:
        plaintext: The data to encrypt
        
    Returns:
        Base64-encoded string containing IV + ciphertext + auth_tag
    """
    # Get the encryption key
    key = get_encryption_key()
    
    # Generate a random 96-bit (12-byte) IV
    iv = secrets.token_bytes(12)
    
    # Create cipher
    cipher = AESGCM(key)
    
    # Encrypt the data
    ciphertext = cipher.encrypt(iv, plaintext.encode('utf-8'), None)
    
    # ciphertext includes the authentication tag at the end (last 16 bytes)
    # Combine IV + ciphertext (which includes auth_tag)
    encrypted_data = iv + ciphertext
    
    # Return base64 encoded result
    return base64.b64encode(encrypted_data).decode('utf-8')

def decrypt_data(encrypted_data: str) -> str:
    """
    Decrypt data encrypted with AES-256-GCM.
    
    Args:
        encrypted_data: Base64-encoded string containing IV + ciphertext + auth_tag
        
    Returns:
        Decrypted plaintext
        
    Raises:
        ValueError: If decryption fails or authentication fails
    """
    # Get the encryption key
    key = get_encryption_key()
    
    # Decode from base64
    encrypted_bytes = base64.b64decode(encrypted_data)
    
    # Extract IV (first 12 bytes)
    iv = encrypted_bytes[:12]
    
    # Extract ciphertext + auth_tag (remaining bytes)
    ciphertext_with_tag = encrypted_bytes[12:]
    
    # Create cipher
    cipher = AESGCM(key)
    
    try:
        # Decrypt and verify authentication tag
        plaintext = cipher.decrypt(iv, ciphertext_with_tag, None)
        return plaintext.decode('utf-8')
    except Exception as e:
        raise ValueError(f"Decryption failed - authentication tag verification failed: {str(e)}")

def generate_encryption_key() -> str:
    """
    Generate a new random 256-bit encryption key.
    
    Returns:
        Hex-encoded 256-bit key suitable for ENCRYPTION_KEY environment variable
    """
    key = secrets.token_bytes(32)  # 32 bytes = 256 bits
    return key.hex()

if __name__ == "__main__":
    # Generate a test key and set it
    test_key = generate_encryption_key()
    os.environ['ENCRYPTION_KEY'] = test_key
    print(f"Generated test key: {test_key}")
    
    # Test encryption and decryption
    plaintext = "Hello, World! This is sensitive data."
    print(f"\nOriginal plaintext: {plaintext}")
    
    # Encrypt
    encrypted = encrypt_data(plaintext)
    print(f"Encrypted (base64): {encrypted}")
    
    # Decrypt
    decrypted = decrypt_data(encrypted)
    print(f"Decrypted plaintext: {decrypted}")
    
    # Verify round-trip
    assert plaintext == decrypted, "Round-trip encryption/decryption failed!"
    print("\n✓ Round-trip encryption/decryption successful!")
    
    # Test with longer data
    long_data = "This is a much longer piece of sensitive data. " * 10
    encrypted_long = encrypt_data(long_data)
    decrypted_long = decrypt_data(encrypted_long)
    assert long_data == decrypted_long, "Long data round-trip failed!"
    print("✓ Long data encryption/decryption successful!")
    
    # Demonstrate that tampering is detected
    print("\nTesting tampering detection...")
    encrypted_clean = encrypt_data("Important data")
    
    # Try to tamper with the encrypted data
    tampered = base64.b64encode(base64.b64decode(encrypted_clean)[:-1] + b'X').decode('utf-8')
    
    try:
        decrypt_data(tampered)
        print("✗ Tampering was NOT detected (this should not happen!)")
    except ValueError as e:
        print(f"✓ Tampering detected: {e}")